portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java

```diff
@@ -966,7 +973,36 @@ public void doLogin(HttpServletRequest req, HttpServletResponse res, Session ses
 	public void doLogout(HttpServletRequest req, HttpServletResponse res,
 			Session session, String returnPath) throws ToolException
 	{
-		
+
+		// if this is an impersonation, then reset the users old session and
+		if(impersonating()){
+			UsageSession oldSession = (UsageSession) session.getAttribute(UsageSessionService.USAGE_SESSION_KEY);
+
+			String userId = oldSession.getUserId();
+			String userEid = oldSession.getUserEid();
+			M_log.warn("returning to user: " + userId + "/" + userEid);
+			//SAK-1245
+			Vector<String> saveAttributes = new Vector<String>();
+			saveAttributes.add(UsageSessionService.USAGE_SESSION_KEY);
+			saveAttributes.add(UsageSessionService.SAKAI_CSRF_SESSION_ATTRIBUTE);
+			session.clearExcept(saveAttributes);
+
+			// login - set the user id and eid into session, and refresh this user's authz information
+			session.setUserId(userId);
+			session.setUserEid(userEid);
+			authzGroupService.refreshUser(userId);
+
+			try {
+				res.sendRedirect(ServerConfigurationService.getString("portalPath", "/portal"));
+				res.getWriter().close();
+			} catch (IOException e) {
+				M_log.error("failed to redirect after Return To Me");
+				e.printStackTrace();
```

Comment:
can add the stacktrace to the M_log.error:

M_log.error("failed to redirect after Return To Me", e);